### ISSUE TITLE
IE11 throws with "Object doesn't support this action"

### ISSUE THREAD

**Issue by nirazul** [ungap/url-search-params#8]
In Internet Explorer 11, the polyfill throws with `Object doesn't support this action` when executing the following line:
`new URLSearchParams("q=URLUtils.searchParams&topic=api");`

According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams#Examples), the call is fine.

I've provided more info in the following issue as I've first suspected Axios:
https://github.com/axios/axios/issues/2045

I've updated from `url-search-params@0.10.0` which was working and stumbled over this bug.

**Comment by WebReflection** [comment]
![Screenshot from 2019-05-09 18-27-40](https://user-images.githubusercontent.com/85749/57470037-3a417d80-7288-11e9-9892-2995c902a71e.png)

**Comment by WebReflection** [comment]
live test page here https://ungap.github.io/url-search-params/test/

**Comment by nirazul** [comment]
@WebReflection 
Maybe I was not explicit enough. The module fails when using it in webpack, I guess that the file used by webpack is not the one you are using in your test. I've created a repo for you to check it:

[url-search-params-test](https://github.com/nirazul/url-search-params-test)

As you can see in IE11, there is no alert and an error is thrown. The only thing I'm doing is:
`window.URLSearchParams = require('@ungap/url-search-params');`

**Comment by WebReflection** [comment]
the CJS file is exactly the same https://github.com/ungap/url-search-params/blob/master/package.json#L10

try requiring it from webpack via `require('@ungap/url-search-params/cjs')` and see if that works

**Comment by nirazul** [comment]
@WebReflection 
Yes, when unsing `@ungap/url-search-params/cjs'` it works! Maybe it's a good idea to change the `package.json` to make it more webpack friendly:
According to [webpack documentation](https://webpack.js.org/configuration/resolve/#resolvemainfields), the default `package.json` fields that are checked are `['browser', 'module', 'main']`. The first one existing is the `module` field. So it defaults to the `esm` build. So you can add the `browser` field with the value `cjs/index.js`.

Would you accept a PR for this?

**Comment by WebReflection** [comment]
> Maybe it's a good idea to change the package.json to make it more webpack friendly

the package json specifies exactly which file CJS bundler should use, through the `main` field.

The fact Webpack does whatever it likes, it's not a personal concern of mine.

Also, IIRC, rollup got this right since day zero.

> According to webpack documentation, the default package.json fields that are checked are ['browser', 'module', 'main']

browser as CJS makes no sense, since no browser ships with CJS. This is an old Webpack gotcha I wish by 2019 they got rid of, but I guess it's their legacy, hopefully fixed as soon as NodeJS modules ship stable, instead of behind a flag.

TL;DR you should file yet another bug in Webpack about this, they should invert the order of their search 'cause `main` **is** the CJS standard, everything else is foot-gun prone.

**Comment by WebReflection** [comment]
P.S. I've updated the readme https://github.com/ungap/url-search-params#-webpack-users

I also didn't explicitly answer to your question:

> Would you accept a PR for this?

No, 'cause this is a Webpack issue 👋

**Comment by nirazul** [comment]
Thanks for editing the readme, that will certainly save some time for anyone who's stumbling over this issue.

P.S.: A personal side note for you:
To me, your answers felt rather aggressive by the tone they are using. I can understand that you might be frustrated or at least annoyed by this topic, but I tried to be as constructive as possible. I can only do so much, check the closed tickets for clues and try to give (hopefully) constructive feedback. I wasn't aware of any debate going on about package.json fields or module types, so please take into consideration that I'm just trying to solve a problem here, ideally for more people than just me. We're both just humans talking about a bunch of lines of text after all, so there's really no reason to be dismissive towards anyone.

**Comment by WebReflection** [comment]
> To me, your answers felt rather aggressive by the tone they are using

apologies for letting you feel this way, the true story is that I'm tired of non standard choices made by popular projects, so that you got the same answer I'd be giving to any webpack contributor.

It wasn't too sensible though, so sorry again for that.

> please take into consideration that I'm just trying to solve a problem here

You are right, I should put more empathy in my quick replies (I'm full time employee, and everything in GitHub is actually (kinda) a side project ... I have dozens of those, actively maintained).

On the other hand, the problem you are trying to solve is from somewhere else, and this is the pattern I keep seeing, but I should probably write a blog post about it.

Every other day, somebody files an issue in one of my projects, because their more hyped libraries/helpers/frameworks have issues, but I get the blame.

When I read things like:

> According to webpack documentation, the default package.json fields that are checked are ['browser', 'module', 'main']

all I understand is: _and why on Earth you believe that is how things should be?_

I have contributed to ECMAScript standards, followed the NodeJS modules history (**very** closely), and yet I have to explain that whatever webpack is doing with the package.json fields is wrong ... now, put this into some perspective:

  * `@ungap/url-search-params` has 80+ stars
  * `webpack` has 48K+ stars

Basically I have 6000% of chances people will bother my libraries for something webpack did wrong, accordingly to standards (in this case the NodeJS CommonJS standard).

Imagine every single time I have people believing the problem is in my projects how I feel ... and yet, I still apology because common developers wouldn't know this, wouldn't bother to think about standards, and I'm somehow one of the few caring about how things should be done, accordingly to standards.

So once again, my apologies, and please don't take this personally, but I'm tired of how Open Source is working these days.

It used to be a cool place to learn, today is a place biased out of popularity, instead of common sense.

**Comment by nirazul** [comment]
Thanks for your openness! I can really understand your frustration about it, especially because of libraries creating new realities out of sheer popularity. I wouldn't have ever proposed changing your `package.json` file if I had known the background of this ;)

Maybe it's a bit silly to give an ECMA contributor such an advice, but one of the few things that help changing realities is educating people. So it's certainly the right call to enclose a note in the readme and I think it's also a good idea to address your concerns in a blog post, so that more developers become aware of the issue.

Anyways, keep up the good work and don't let frustration eat you up, it's not worth it 👍

**Comment by mornir** [comment]
Oh! This is why my axios post requests were failing in IE11 😮 I thought the problem was coming from axios. I'm glad I found this thread!

**Comment by WebReflection** [comment]
@mornir you should also read answers, as there is no IE11 error with this library, as you can test already live: https://ungap.github.io/url-search-params/test/

**Comment by ghost** [comment]
@WebReflection This views on standards are ridiculous. I just needed something to make it work. This just doesn't work with webpack even if i import from cjs path. 
Just delete your library if you hate open-source so much. I would have written myself rather than wasting my time with some broken third party. 
There is nothing standard about npm. And if you obey standards you could not move far.

**Comment by WebReflection** [comment]
@atilkan thanks for your super constructive feedback, the Open Source community definitively needs more people like you. I have already reported your kind comment, and I'll wait a reply before blocking you from my repositories, so you'll have all the freedom to rewrite all the polyfills by your own, as I am sure you'll do a better job than anyone else.

Best Regards.

**Comment by WebReflection** [comment]
P.S.

> This just doesn't work with webpack even if i import from cjs path.

That's because both [me](https://github.com/ungap/url-search-params#-webpack-users) and [nirazul](https://github.com/ungap/url-search-params/issues/8#issuecomment-491257245) lied only to make you waste your precious time.
